interfaces/builtin/custom_device.go

```diff
@@ -0,0 +1,394 @@
+	customDeviceAttr, isSet := plug.Attrs["custom-device"]
+	customDevice, ok := customDeviceAttr.(string)
+	if isSet && !ok {
+		return fmt.Errorf(`custom-device "custom-device" attribute must be a string, not %v`,
+			plug.Attrs["custom-device"])
+	}
+	if customDevice == "" {
+		if plug.Attrs == nil {
+			plug.Attrs = make(map[string]interface{})
+		}
+		// custom-device defaults to "plug" name if unspecified
+		plug.Attrs["custom-device"] = plug.Name
+	}
+
+	return nil
+}
+
+func (iface *customDeviceInterface) AppArmorConnectedPlug(spec *apparmor.Specification, plug *interfaces.ConnectedPlug, slot *interfaces.ConnectedSlot) error {
+	snippet := &bytes.Buffer{}
+	emitRule := func(paths []string, permissions string) {
+		for _, path := range paths {
+			fmt.Fprintf(snippet, "\"%s\" %s,\n", path, permissions)
+		}
+	}
+
+	// get all attributes without validation, since that was done before;
+	// should an error occur, we'll simply not write any rule.
+
+	var devicePaths []string
+	_ = slot.Attr("devices", &devicePaths)
+	emitRule(devicePaths, "rw")
+
+	var readDevicePaths []string
+	_ = slot.Attr("read-devices", &readDevicePaths)
+	emitRule(readDevicePaths, "r")
+
+	var writablePaths []string
+	_ = slot.Attr("write", &writablePaths)
+	emitRule(writablePaths, "rw")
+
+	var readOnlyPaths []string
+	_ = slot.Attr("read", &readOnlyPaths)
+	emitRule(readOnlyPaths, "r")
+
+	spec.AddSnippet(snippet.String())
+	return nil
+}
+
+// extractStringMapAttribute looks up the given key in the container, and
+// returns its value as a map[string]string.
+// No validation is performed, since it already occurred before connecting the
+// interface.
+func (iface *customDeviceInterface) extractStringMapAttribute(container map[string]interface{}, key string) map[string]string {
+	valueMap, ok := container[key].(map[string]interface{})
+	if !ok {
+		return nil
+	}
+
+	stringMap := make(map[string]string, len(valueMap))
+	for key, value := range valueMap {
+		stringMap[key] = value.(string)
+	}
+
+	return stringMap
+}
+
+func (iface *customDeviceInterface) UDevConnectedPlug(spec *udev.Specification, plug *interfaces.ConnectedPlug, slot *interfaces.ConnectedSlot) error {
+	// Collect all the device paths specified in either the "devices" or
+	// "read-devices" attributes.
+	var devicePaths []string
+	_ = slot.Attr("devices", &devicePaths)
+	var readDevicePaths []string
+	_ = slot.Attr("read-devices", &readDevicePaths)
+	allDevicePaths := devicePaths
+	allDevicePaths = append(allDevicePaths, readDevicePaths...)
+
+	// Generate a basic udev rule for each device; we put them into a map
+	// indexed by the device name, so that we can overwrite the entry later
+	// with a more specific rule.
+	deviceRules := make(map[string]string, len(allDevicePaths))
+	for _, devicePath := range allDevicePaths {
+		if strings.HasPrefix(devicePath, "/dev/") {
+			deviceName := devicePath[5:]
+			deviceRule := fmt.Sprintf(`KERNEL=="%s"`, deviceName)
+			deviceRules[deviceName] = deviceRule
```

Comment:
I think it's slightly more readable if separate (and more consistent since I also declared a variable for `deviceName`), but I've got no strong opinion on this. I'll apply your suggestion.